test/integration/scheduler/volume_equiv_cache_test.go

```diff
@@ -0,0 +1,435 @@
+			equivPodShouldFail: false,
+		},
+		"On PV region update": {
+			pod:  makePod("pod-pvc-1", config.ns, []string{"pvc-canbind-1"}),
+			pvs:  []*v1.PersistentVolume{pv},
+			pvcs: []*v1.PersistentVolumeClaim{makePVC("pvc-canbind-1", config.ns, &classImmediate, pv.Name)},
+			action: func() {
+				updatePVRegion(t, pv.Name, config)
+			},
+			equivPod: makePod("pod-2-pvc-1", config.ns, []string{"pvc-canbind-1"}),
+			// Scheduling should fail since PV has been updated with a invalid region.
+			equivPodShouldFail: true,
+		},
+	}
+
+	runTest(t, cases, config)
+}
+
+// TestMaxGCEPDVolumeCountPred verifies MaxGCEPDVolumeCountPred will be properly invalidated after certain actions happen.
+func TestMaxGCEPDVolumeCountPred(t *testing.T) {
+	// Setup a cluster with max PD volume count to 1.
+	os.Setenv(predicates.KubeMaxPDVols, strconv.Itoa(1))
+
+	policy := &schedulerapi.Policy{
+		Predicates: []schedulerapi.PredicatePolicy{
+			{
+				Name: predicates.MaxGCEPDVolumeCountPred,
+			},
+		},
+	}
+	policy.APIVersion = testapi.Groups[v1.GroupName].GroupVersion().String()
+	config := setupClusterWithOptions(t, "volume-scheduling", 1, policy, "")
+	defer config.teardown()
+
+	podPv1 := makePod("pod-pv1", config.ns, []string{})
+	podPv1.Spec.Volumes = []v1.Volume{
+		{
+			Name: "pv-gcepd-2",
+			VolumeSource: v1.VolumeSource{
+				GCEPersistentDisk: &v1.GCEPersistentDiskVolumeSource{
+					PDName: "test-gcepd-2",
+				},
+			},
+		},
+	}
+
+	podPv2 := makePod("pod-pv2", config.ns, []string{})
+	podPv2.Spec.Volumes = []v1.Volume{
+		{
+			Name: "pv-gcepd-2",
+			VolumeSource: v1.VolumeSource{
+				GCEPersistentDisk: &v1.GCEPersistentDiskVolumeSource{
+					PDName: "test-gcepd-2",
+				},
+			},
+		},
+	}
+
+	podPv := makePod("pod-pv", config.ns, []string{})
+	podPv.Spec.Volumes = []v1.Volume{
+		{
+			Name: "pv-gcepd",
+			VolumeSource: v1.VolumeSource{
+				GCEPersistentDisk: &v1.GCEPersistentDiskVolumeSource{
+					PDName: "test-gcepd",
+				},
+			},
+		},
+	}
+
+	// These test cases will create work-flow as below:
+	// 1. Create first Pod, its schedule result is cached in ecache.
+	// 2. Perform action to change the status of MaxGCEPDVolumeCountPred, i.e. abuse/release PD count limit.
+	// 3. Create the equivalent Pod of first Pod, it should have opposite schedule result, since it is re-calculated with
+	//    respect to step 2. We need to create opposite schedule result to verify ecache data is not reused.
+	cases := testCase{
+		"On Pod delete": {
+			existingPods: []*v1.Pod{podPv},
+			pod:          podPv1,
+			// Fail because PD count limit is abused
+			firstShouldFail: true,
+			action: func() {
+				removePod(t, podPv.Name, config)
+			},
+			// The equivalent Pod should not fail since PD count limit is released by Pod deletion.
+			equivPod:           podPv2,
+			equivPodShouldFail: false,
+		},
+		// TODO(harry): it's not clear how to test "On Pod add" because a equivalent Pod always has the same PVC
```

Comment:
What about this case:
1) Pod A passes volume count pred, but fails scheduling due to some other predicate
2) Pod B comes along which also uses the same volume type and does get scheduled successfully
3) Now Pod A cannot get scheduled because Pod B used up the volume count limit

Reply:
Emm ... I think in step 3, we will want to check an equivalent pod of A, let's say, A-E, and assume A is failed because of predicate X. 

Then we need to make sure A-E 's failure of scheduling is not caused by X. It's still tricky, as none of the existing predicate behaves differently for A and A-E ... unless, we set a custom predicate func, or, check the failure message instead (not sure how). WDYT?

Reply:
A scenario I had in mind could be something like:
1. Pod A uses volume type X and has some nodeSelector.  It passed the volume count predicate, but failed the nodeSelector predicate.
2. Pod B uses volume type X but does not have a nodeSelector.  It passes the volume count predicate, and gets scheduled.
4. Add a label to the node so that the nodeSelector passes.
3. Pod A should fail the volume count predicate now.

Is the challenge you are describing that the test currently only checks success/failure of scheduling in general, but not specific predicates?

Reply:
Also, to clarify here, Pod B can use a different PVC/PV but using the same volume type, ie gce pd.  So Pod A and B are not equivalent.

Reply:
> Is the challenge you are describing that the test currently only checks success/failure of scheduling in general, but not specific predicates?

Yes, exactly. 

So, maybe we can try to check the error message of pod scheduling failure.